plugins/logitech-bulkcontroller/fu-logitech-bulkcontroller-device.c

```diff
@@ -423,6 +423,145 @@ fu_logitech_bulkcontroller_device_compute_hash(GBytes *data)
+	if (json_device == NULL) {
+		g_set_error_literal(error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_INVALID_DATA,
+				    "did not get JSON device");
+		return FALSE;
+	}
+	if (json_object_has_member(json_device, "name"))
+		fu_device_set_name(device, json_object_get_string_member(json_device, "name"));
+	if (json_object_has_member(json_device, "sw"))
+		fu_device_set_version(device, json_object_get_string_member(json_device, "sw"));
+	if (json_object_has_member(json_device, "type"))
+		fu_device_add_instance_id(device,
+					  json_object_get_string_member(json_device, "type"));
+	if (json_object_has_member(json_device, "status"))
+		self->status = json_object_get_int_member(json_device, "status");
+	if (json_object_has_member(json_device, "updateStatus"))
+		self->update_status = json_object_get_int_member(json_device, "updateStatus");
+
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_bulkcontroller_device_get_data(FuDevice *device, GError **error)
+{
+	FuLogitechBulkcontrollerDevice *self = FU_LOGITECH_BULKCONTROLLER_DEVICE(device);
+	g_autoptr(GByteArray) device_request = g_byte_array_new();
+	g_autoptr(GByteArray) decoded_pkt = g_byte_array_new();
+	g_autoptr(GByteArray) device_response = g_byte_array_new();
+	FuLogitechBulkcontrollerProtoId proto_id = kProtoId_UnknownId;
+
+	/* sending GetDeviceInfoRequest. Device reports quite a few matrix, including status,
+	 * progress etc */
+	device_request = proto_manager_generate_get_device_info_request();
+	if (!fu_logitech_bulkcontroller_device_send_sync_cmd(self,
+							     CMD_BUFFER_WRITE,
+							     device_request,
+							     error)) {
+		g_prefix_error(error,
+			       "error in sending buffer write packet for device info request: ");
+		return FALSE;
+	}
+
+	/* wait for the GetDeviceInfoResponse */
+	if (!fu_logitech_bulkcontroller_device_recv_sync_cmd(self,
+							     CMD_BUFFER_READ,
+							     device_response,
+							     error)) {
+		g_prefix_error(error, "error in buffer read packet for device info request: ");
+		return FALSE;
+	}
+	if (!fu_logitech_bulkcontroller_device_recv_sync_cmd(self,
+							     CMD_UNINIT_BUFFER,
+							     NULL,
+							     error)) {
+		g_prefix_error(error, "error in buffer read packet for device info request: ");
+		return FALSE;
+	}
+	if (!fu_logitech_bulkcontroller_device_send_sync_cmd(self,
+							     CMD_UNINIT_BUFFER,
+							     NULL,
+							     error)) {
+		g_prefix_error(
+		    error,
+		    "error in sending buffer uninitialize packet for device info request: ");
+		return FALSE;
+	}
+	decoded_pkt = proto_manager_decode_message(device_response->data,
+						   device_response->len,
+						   &proto_id,
+						   error);
+	if (decoded_pkt == NULL) {
+		g_prefix_error(error, "error in unpacking packet for device info request: ");
+		return FALSE;
+	}
+	g_debug("Received device response. data: %s, length: %u",
+		(const gchar *)decoded_pkt->data,
+		decoded_pkt->len);
```

Comment:
printing this as text could be unsafe and probably overkill to put in the logs every time. I think you want something like:

	if (g_getenv("FWUPD_LOGITECH_BULKCONTROLLER_VERBOSE") != NULL)
		fu_common_dump_raw(G_LOG_DOMAIN, "Response", decoded_pkt->data, decoded_pkt->len);

Reply:
Fixed. fu_common_dump_raw is not practical at this time

Reply:
Is it because you want to print the JSON? if so, you need to use fu_common_strsafe() -- You don't know if fu_common_strsafe is NUL terminated or not.

Reply:
Good point. Fixed